### PR TITLE
Add user-level settings min_free_diskspace_bytes_to_throw_insert and min_free_diskspace_ratio_to_throw_insert

### LINKED ISSUE #65951: Add setting to stop INSERTs when disk space is close to 100%
**Use Case**

When ClickHouse fills the disk to 100%, it creates several negative side effects: inability to perform certain SELECT queries (which require temporary data), potential file corruption, inability to write logs, and other cascading issues that complicate recovery. It's preferable to stop writes slightly earlier to avoid these problems.

The same logic as `parts_to_throw_insert` can be applied here. Eventually, writes will fail as the disk becomes full; it's better if they fail slightly earlier to prevent cascading problems.

**Describe the solution you'd like**

Introduce user-level settings: `min_free_diskspace_bytes_to_throw_insert ` (default 10GB) and `min_free_diskspace_ratio_to_throw_insert ` (default 0.02, i.e., 2%). When these settings are non-zero, during an INSERT operation, ClickHouse will ensure that the storage policy of the target table has at least `min_free_diskspace_bytes_to_throw_insert` bytes or `min_free_diskspace_ratio_to_throw_insert` free space available. If not, it will throw an exception.

**Describe alternatives you've considered**

Relying on user discipline, monitoring, and alerting has proven to be ineffective.

**Additional context**

Throttling / delaying / slowing down the inserts in that conditions is more controversial, so I don't think it's really needed.

May be we can also make the logic a bit more complex - by checking if the default drive have that amount of free disks space AND there is at least one disk in storage policy of the table with that amount of the free disk space.

### PR DESCRIPTION
**Pull request by marco-vb**
Resolves #65951 

Added two new user-level settings:
- `min_free_disk_bytes_to_throw_insert`
- `min_free_disk_ratio_to_throw_insert`

These prevent inserts from occurring if disk is close to full, but do not affect operations that need to write temporary data to disk. See #65951 discussion to better understand why this is useful.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added user-level settings `min_free_disk_bytes_to_throw_insert` and `min_free_disk_ratio_to_throw_insert` to prevent insertions on disks that are almost full.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 4cbe85ad1535a49d51bdac7fd05d154cdcb18f82 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69755/4cbe85ad1535a49d51bdac7fd05d154cdcb18f82/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69755/4cbe85ad1535a49d51bdac7fd05d154cdcb18f82/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by marco-vb** [comment]
@yariks5s thanks for adding test tag, I think this should be ok now even though one build got cancelled for some reason

File: src/Storages/MergeTree/MergeTreeDataWriter.cpp
Comment (yariks5s) [reply src/Storages/MergeTree/MergeTreeDataWriter.cpp:590]: or even better: `"Could not perform insert: less than {} free bytes left in the disk space ({}).`

File: src/Core/Settings.cpp
Comment (marco-vb) [reply src/Core/Settings.cpp:363]: Yes, I have been using this name because I was guiding myself by what was originally described in the issue #65951 but having the 'throw' word (even in the setting) can be misleading as it may cause some confusion on wether the insert is performed or an exception is thrown. Will change the setting name as well

File: src/Storages/MergeTree/MergeTreeSettings.cpp
Comment (yariks5s) [reply src/Storages/MergeTree/MergeTreeSettings.cpp:235]: Let's suppose we have some unusual value for this setting; we pass the sanity check, and now we have a default value for this setting, how do we catch this exception then?

Comment (marco-vb) [reply src/Storages/MergeTree/MergeTreeSettings.cpp:235]: From what I have been testing, it seems that even if we pass the first sanity check (the one in `SettingsQuirks.cpp` we can still throw this exception: let's say we have default `min_free_disk_ratio_to_throw_insert = 0.0` and then we send to the server the query `create table b (id UInt64) engine = MergeTree order by id settings min_free_disk_ratio_to_perform_insert = 100`, then this query will fail with this exception and the table will not be created.

So basically my approach was to guarantee that there is always a default 'safe' value in the server config, and also prevent a individual table level to have an invalid value. Please let me know if this is reasonable or if I am just overcomplicating things, since I am quite new to this project.

File: src/Storages/MergeTree/MergeTreeDataWriter.cpp
Comment (marco-vb) [reply src/Storages/MergeTree/MergeTreeDataWriter.cpp:571]: I will test this more thoroughly but I think I need to have this to make sure that the most granular level setting is used -- that is, let's say I just use the value from `data_settings` and I create a table with `create table a (id UInt64) engine = MergeTree order by id`, this table `a` will have the default value of `min_free_disk_ratio_to_throw_insert = 0.0` as defined in `MergeTreeSettings.h`. Which means that even if we have a server config of `min_free_disk_ratio_to_throw_insert = 0.05`, the table `a` will never see this value and therefore keep accepting inserts until the disk is completely full.

Also please take a look at #69376 -- this is a previous attempt of mine to implement this setting and reviewers point out the fact that even though the setting is defined both in `Core` and `MergeTreeSettings` only the latter value is ever checked.

File: src/Storages/MergeTree/MergeTreeSettings.cpp
Comment (yariks5s) [reply src/Storages/MergeTree/MergeTreeSettings.cpp:235]: I think it will be better just to throw an exception without a sanity check because, as you said, if you get an exception if the data is incorrect, then a sanity check does not much sense🙂

Comment (marco-vb) [reply src/Storages/MergeTree/MergeTreeSettings.cpp:235]: Ok I see, does make sense thanks!

**Comment by marco-vb** [comment]
hey @yariks5s thank you very much for the detailed review and for insisting on the CI/CD -- for some reason it keeps cancelling the `Builds_2/binary_tidy` and therefore giving failed status to builds check... In the meantime I fixed a merge conflict in the last commit and now the CI/CD seems to need approval again

**Comment by yariks5s** [comment]
Hi! Sorry for the delay, binary_tidy was broken some time ago. I will take a look at the code once more, and if it's OK, we will merge it soon.

File: docs/en/operations/settings/merge-tree-settings.md
Comment (yariks5s) [inline docs/en/operations/settings/merge-tree-settings.md:1088]: I'd add here something like this:

Note that if both `min_free_disk_ratio_to_perform_insert` and `min_free_disk_bytes_to_perform_insert` are specified, ClickHouse will count on the value that will allow to perform inserts on a bigger amount of free memory.

**Review by yariks5s — APPROVED** [review]
Generally LGTM👍

File: docs/en/operations/settings/merge-tree-settings.md
Comment (marco-vb) [reply docs/en/operations/settings/merge-tree-settings.md:1088]: Agreed, updated!

File: src/Storages/MergeTree/MergeTreeDataWriter.cpp
Comment (marco-vb) [reply src/Storages/MergeTree/MergeTreeDataWriter.cpp]: Had ratio in mind for some reason thanks for pointing out.

**Review by yariks5s — APPROVED** [review]
LGTM, thank you!🙂 Let's wait for the tests

File: docs/en/operations/settings/merge-tree-settings.md
Comment (marco-vb) [reply docs/en/operations/settings/merge-tree-settings.md:1095]: Yes this makes sense, will make a PR soon.

**Pull request merged**
